realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java

```diff
@@ -204,6 +204,14 @@ public RealmObjectSchema removeField(String fieldName) {
         if (table.getPrimaryKey() == columnIndex) {
             table.setPrimaryKey(null);
         }
+        // When PK does not exist (-2) this cannot happen. Meanwhile a removal target has a smaller
+        // index than the PK index, we can expect an rearrangement.
+        if (columnIndex < table.getPrimaryKey()) {
+            String pkField = getPrimaryKey();
+            table.removeColumn(columnIndex);
+            table.setPrimaryKey(pkField);
```

Comment:
Reading through the Table class.  This seems like a hack to recalculate the `cachedPrimaryKeyColumnIndex` in the `Table` class. Wouldn't it be easier to have a `invalidatePrimaryKeyIndex()` method instead that could set the new proper index?